### ISSUE TITLE
how well would this work with 10M documents?

### ISSUE THREAD

**Issue by chovy** [louischatriot/nedb#476]
documents are only a few kilobytes.

You mentioned expected datasets are 10k -- but I plan on storing 10M on the upper limit.

If this isn't going to work, do you recommend a similar javascript-ish document store?

**Comment by jampy** [comment]
You mean 10,000,000 documents x a few kilobytes?

I guess that is going to be difficult and would take quite some time to load.

**Comment by chovy** [comment]
Yeah. I'm testing it out now. At 20k docs so far. Will take a few days to generate 1M

**Comment by chovy** [comment]
I hit the 256MB database size limit at around 100k documents. Its too bad because it seems to work just fine right up until I hit the 256MB size.

**Comment by jampy** [comment]
see #389

**Comment by chovy** [comment]
I have 72GB of memory. 256MB is nothing.

**Comment by tsturzl** [comment]
Then have at it, but index generation is going to mean loading the entire dataset into memory every time you start the application. Its also worth noting that nodejs by default has a maximum memory limit(I believe 1GB on 64bit systems for latest LTS), you can set it higher with a parameter(this might be your issue). The memory to index the document might also be higher than that, because its stored in an AVL Tree. Either way you slice it, you're going to have a high startup overhead that just grows exponentially as your dataset grows. Though it seems you know your upper limit already, so that can help you assess the startup overhead. I'd say loading 258mb into memory shouldn't take too long, but I don't have enough information to tell you how long it would take to generate the indices because ALV Tree's do not have a fast insertion speed because they are self balancing tree structures, so it would mean a lot of overhead to balance the tree while generating the indices. I know index generation is incredibly CPU intensive on heavy inserts and on startup, so thats something to bare in mind also.

Something lower level might serve you better, and thats something to keep in mind while making your decision. Sorry I jumped to a conclusion. These are all things you'll need to asses on your end.

**Comment by chovy** [comment]
I can't use more than 256MB because NeDB hardcodes that as max database size.

**Comment by tsturzl** [comment]
Sure enough, but even then there are probably still better options. I should have read the issue jampy referenced.

**Comment by chovy** [comment]
I agree. I was just going to see how much it could reasonably support but I was capped at 256MB and it was performing fine.

I've moved to rethinkdb.
